### ISSUE TITLE
Possible bug with assert()

### ISSUE THREAD

**Issue by JamesTheHacker** [Respect/Validation#1180]
The bug has already been explained here: https://stackoverflow.com/questions/54351837/unable-to-catch-exception-from-assert-method-using-respect-valiation

Here is the offending code: https://github.com/Respect/Validation/blob/1.1/library/Rules/KeySet.php#L121

If an error occurs in `checkKeys` when called from `assert` the exception that is thrown in `checkKeys` is not caught or rethrown in `assert`. This causes the unhandled exception in the SO. Here's the line from `KeySet.php`:

     public function assert($input)
     {
        $this->checkKeys($input);
        return parent::assert($input);
     }

**Comment by JamesTheHacker** [comment]
My bad. I'm wrong. As can be seen here:

```
function a() {
    throw new Exception("Woops");
}

function b() {
    a();
}

function c() {
    b();
}

try {
    c();
} catch (Exception $e) {
    print("Oh no!");
}
```
